internal/databases/greenplum/dir_database_tar_ball_composer.go

```diff
@@ -0,0 +1,291 @@
+
+func (d *DirDatabaseTarBallComposer) AddFile(info *internal.ComposeFileInfo) {
+	select {
+	case d.addFileQueue <- info:
+		return
+	case <-d.ctx.Done():
+		tracelog.ErrorLogger.Printf("AddFile: not doing anything, err: %v", d.ctx.Err())
+		return
+	}
+}
+
+func (d *DirDatabaseTarBallComposer) AddHeader(fileInfoHeader *tar.Header, info os.FileInfo) error {
+	tarBall, err := d.tarBallQueue.DequeCtx(d.ctx)
+	if err != nil {
+		return d.errorGroup.Wait()
+	}
+	tarBall.SetUp(d.crypter)
+	defer d.tarBallQueue.EnqueueBack(tarBall)
+	d.tarFileSetsMutex.Lock()
+	d.tarFileSets.AddFile(tarBall.Name(), fileInfoHeader.Name)
+	d.tarFileSetsMutex.Unlock()
+	d.files.AddFile(fileInfoHeader, info, false)
+	return tarBall.TarWriter().WriteHeader(fileInfoHeader)
+}
+
+func (d *DirDatabaseTarBallComposer) SkipFile(tarHeader *tar.Header, fileInfo os.FileInfo) {
+	d.files.AddSkippedFile(tarHeader, fileInfo)
+}
+
+func (d *DirDatabaseTarBallComposer) FinishComposing() (internal.TarFileSets, error) {
+	close(d.addFileQueue)
+
+	err := d.errorGroup.Wait()
+	if err != nil {
+		return nil, err
+	}
+
+	d.addFileWaitGroup.Wait()
+
+	err = internal.UploadDto(d.uploader.Folder(), d.aoStorageUploader.GetFiles(), getAOFilesMetadataPath(d.backupName))
+	if err != nil {
+		return nil, fmt.Errorf("failed to upload AO files metadata: %v", err)
+	}
+
+	// Push Headers in first part
+	err = d.addListToTar(make([]*internal.ComposeFileInfo, 0))
+	if err != nil {
+		return nil, err
+	}
+
+	eg := errgroup.Group{}
+	for _, fileInfos := range d.fileDirCollection {
+		thisInfos := fileInfos
+		eg.Go(func() error {
+			if len(thisInfos) == 0 {
+				return nil
+			}
+			return d.addListToTar(thisInfos)
+		})
+	}
+
+	if err := eg.Wait(); err != nil {
+		return nil, err
+	}
+
+	return d.tarFileSets, nil
+}
+
+func (d *DirDatabaseTarBallComposer) addListToTar(files []*internal.ComposeFileInfo) error {
+	tarBall := d.tarBallQueue.Deque()
+	tarBall.SetUp(d.crypter)
+
+	for _, file := range files {
+		d.tarFileSets.AddFile(tarBall.Name(), file.Header.Name)
+		err := d.tarFilePacker.PackFileIntoTar(file, tarBall)
+		if err != nil {
+			return err
+		}
+
+		if tarBall.Size() > d.tarBallQueue.TarSizeThreshold {
+			err := d.tarBallQueue.FinishTarBall(tarBall)
+			if err != nil {
+				return err
+			}
+			tarBall = d.tarBallQueue.Deque()
+			tarBall.SetUp(d.crypter)
+		}
+	}
+	return d.tarBallQueue.FinishTarBall(tarBall)
+}
+
+func (d *DirDatabaseTarBallComposer) GetFiles() internal.BundleFiles {
+	return d.files
+}
+
+func (d *DirDatabaseTarBallComposer) addFileWorker(tasks <-chan *internal.ComposeFileInfo) error {
+	for task := range tasks {
+		err := d.addFile(task)
+		if err != nil {
+			tracelog.ErrorLogger.Printf(
+				"Received an error while adding the file %s: %v", task.Path, err)
+			return err
+		}
+	}
+	d.addFileWaitGroup.Done()
+	return nil
+}
+
+func (d *DirDatabaseTarBallComposer) addFile(cfi *internal.ComposeFileInfo) error {
+	// WAL-G uploads AO/AOCS relfiles to a different location
```

Comment:
This logic is getting duplicated. Maybe extract it to some common place? And maybe rename `greenplum_tar_ball_composer` to the `regular_tar_ball_composer` since it is no longer the only one?